actionpack/lib/action_dispatch/journey/formatter.rb

```diff
@@ -33,6 +33,7 @@ def generate(name, options, path_parameters, parameterize = nil)
           defaults       = route.defaults
           required_parts = route.required_parts
           parameterized_parts.delete_if do |key, value|
+            next if defaults[key].nil?
```

Comment:
I haven't really considered its merits (or checked it's right :wink:), but here's a less negative spelling of that:

```ruby
parameterized_parts.keep_if do |key, value|
  defaults[key].nil? ||
    value.to_s != defaults[key].to_s ||
    required_parts.include?(key)
end
```